File: tests/src/test/java/com/optivem/atddaccelerator/templategenerator/systemtest/dsl/github/helpers/ReadmeClient.java
package com.optivem.atddaccelerator.templategenerator.systemtest.dsl.github.helpers;

import com.optivem.atddaccelerator.templategenerator.systemtest.clients.GithubClient;
import com.optivem.atddaccelerator.templategenerator.systemtest.util.Constants;
import com.optivem.atddaccelerator.templategenerator.systemtest.util.Language;

import static org.assertj.core.api.Assertions.assertThat;

public class ReadmeClient {

    private static final String README_PATH = "README.md";

    private final GithubClient client;
    private final String repositoryPath;

    public ReadmeClient (GithubClient client) {
        this.client = client;
        this.repositoryPath = client.getRepositoryPath();
    }

    public void verifyReadmeHasBadges(Language systemLanguage, Language systemTestLanguage) {
        verifyReadmePagesBadge();
        verifyReadmeStageLanguageBadge(Constants.COMMIT_STAGE_MONOLITH_FORMAT, systemLanguage);
        verifyReadmeStageLanguageBadge(Constants.LOCAL_ACCEPTANCE_STAGE_TEST_FORMAT, systemTestLanguage);
        verifyReadmeStageLanguageBadge(Constants.ACCEPTANCE_STAGE_TEST_FORMAT, systemTestLanguage);
        verifyReadmeStageLanguageBadge(Constants.QA_STAGE_TEST_FORMAT, systemTestLanguage);
        verifyReadmeStageLanguageBadge(Constants.PROD_STAGE_TEST_FORMAT, systemTestLanguage);
    }

    private void verifyReadmePagesBadge() {
        var badgeWorkflow = String.format(Constants.PAGES_BUILD_DEPLOYMENT_WORKFLOW_FORMAT, repositoryPath);
        var badgeSvg = String.format(Constants.PAGES_BUILD_DEPLOYMENT_WORKFLOW_IMAGE_FORMAT, repositoryPath);
        verifyReadmeContainsBadge(Constants.PAGES_BUILD_DEPLOYMENT, badgeWorkflow, badgeSvg);
    }

    private void verifyReadmeStageLanguageBadge(String workflowNameFormat, Language language){
        for(var l : Language.getAll()) {
            var workflowName = String.format(workflowNameFormat, l.getValue());
            if(l.equals(language)) {
                verifyReadmeContainsBadge(workflowName);
            } else {
                verifyReadmeDoesNotContainBadge(workflowName);
            }
        }
    }

    private void verifyReadmeContainsBadge(String workflowName) {
        var badgeWorkflow = String.format(Constants.STAGE_WORKFLOW_FORMAT, repositoryPath, workflowName);
        var badgeSvg = String.format(Constants.STAGE_WORKFLOW_IMAGE_FORMAT, repositoryPath, workflowName);

        verifyReadmeContainsBadge(workflowName, badgeWorkflow, badgeSvg);
    }

    private void verifyReadmeContainsBadge(String badgeName, String badgeWorkflow, String badgeSvg) {
        var readmeContent = getReadmeContent();

        assertThat(readmeContent)
                .as("README should contain badge: " + badgeName)
                .contains(badgeName);

        assertThat(readmeContent)
                .as("README should contain badge workflow link: " + badgeWorkflow)
                .contains(badgeWorkflow);

        assertThat(readmeContent)
                .as("README should contain badge SVG: " + badgeSvg)
                .contains(badgeSvg);
    }

    private void verifyReadmeDoesNotContainBadge(String badge) {
        var readmeContent = getReadmeContent();
        assertThat(readmeContent)
                .as("README should not contain badge: " + badge)
                .doesNotContain(badge);
    }

    private String getReadmeContent() {
        return client.getFileContent(README_PATH);
    }

}

File: console/tests/Dsl/GitHub/Helpers/ReadmeClient.cs
using Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Clients;
using Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Util;
using FluentAssertions;

namespace Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Dsl.GitHub.Helpers
{
    public class ReadmeClient
    {
        private const string ReadmePath = "README.md";

        private readonly GithubClient _client;
        private readonly string _repositoryPath;

        public ReadmeClient(GithubClient client)
        {
            _client = client;
            _repositoryPath = client.GetRepositoryPath();
        }

        public void VerifyReadmeHasBadges(Language systemLanguage, Language systemTestLanguage)
        {
            VerifyReadmePagesBadge();
            VerifyReadmeStageLanguageBadge(Constants.CommitStageMonolithFormat, systemLanguage);
            VerifyReadmeStageLanguageBadge(Constants.LocalAcceptanceStageTestFormat, systemTestLanguage);
            VerifyReadmeStageLanguageBadge(Constants.AcceptanceStageTestFormat, systemTestLanguage);
            VerifyReadmeStageLanguageBadge(Constants.QaStageTestFormat, systemTestLanguage);
            VerifyReadmeStageLanguageBadge(Constants.ProdStageTestFormat, systemTestLanguage);
        }

        private void VerifyReadmePagesBadge()
        {
            var badgeWorkflow = string.Format(Constants.PagesBuilderDeploymentWorkflowFormat, _repositoryPath);
            var badgeSvg = string.Format(Constants.PagesBuilderDeploymentWorkflowImageFormat, _repositoryPath);
            VerifyReadmeContainsBadge(Constants.PagesBuilderDeployment, badgeWorkflow, badgeSvg);
        }

        private void VerifyReadmeStageLanguageBadge(string workflowNameFormat, Language language)
        {
            foreach (var l in LanguageExtensions.GetAll())
            {
                var workflowName = string.Format(workflowNameFormat, l.GetValue());
                if (l.Equals(language))
                {
                    VerifyReadmeContainsBadge(workflowName);
                }
                else
                {
                    VerifyReadmeDoesNotContainBadge(workflowName);
                }
            }
        }

        private void VerifyReadmeContainsBadge(string workflowName)
        {
            var badgeWorkflow = string.Format(Constants.StageWorkflowFormat, _repositoryPath, workflowName);
            var badgeSvg = string.Format(Constants.StageWorkflowImageFormat, _repositoryPath, workflowName);

            VerifyReadmeContainsBadge(workflowName, badgeWorkflow, badgeSvg);
        }

        private void VerifyReadmeContainsBadge(string badgeName, string badgeWorkflow, string badgeSvg)
        {
            var readmeContent = GetReadmeContent();

            readmeContent.Should().Contain(badgeName, $"Expected README to contain badge name '{badgeName}', but it was not found.");
            readmeContent.Should().Contain(badgeWorkflow, $"Expected README to contain badge workflow '{badgeWorkflow}', but it was not found.");
            readmeContent.Should().Contain(badgeSvg, $"Expected README to contain badge SVG '{badgeSvg}', but it was not found.");
        }

        private void VerifyReadmeDoesNotContainBadge(string badge)
        {
            var readmeContent = GetReadmeContent();
            readmeContent.Should().NotContain(badge, $"Expected README to NOT contain badge '{badge}', but it was found.");
        }

        private string GetReadmeContent()
        {
            return _client.GetFileContent(ReadmePath);
        }
    }
}